File: java/lib/src/main/java/io/zrz/httpsig/KeyPair.java
package io.zrz.httpsig;

import java.util.Objects;

/**
 * Bundles a {@link SigningKey} and {@link VerifyingKey} that share the same key ID and algorithm.
 */
public record KeyPair(SigningKey signingKey, VerifyingKey verifyingKey) {

    public KeyPair {
        Objects.requireNonNull(signingKey);
        Objects.requireNonNull(verifyingKey);
        if (!signingKey.keyId().equals(verifyingKey.keyId())) {
            throw new IllegalArgumentException("signing and verifying key IDs must match");
        }
        if (signingKey.algorithm() != verifyingKey.algorithm()) {
            throw new IllegalArgumentException("signing and verifying algorithms must match");
        }
    }

    /** The key ID shared by both halves. */
    public String keyId() {
        return signingKey.keyId();
    }

    /** The algorithm shared by both halves. */
    public Algorithm algorithm() {
        return signingKey.algorithm();
    }
}

File: kotlin/lib/src/main/kotlin/io/zrz/httpsig/KeyPair.kt
package io.zrz.httpsig

/**
 * Bundles a [SigningKey] and [VerifyingKey] that share the same key ID and algorithm.
 */
data class KeyPair(
    val signingKey: SigningKey,
    val verifyingKey: VerifyingKey,
) {
    init {
        require(signingKey.keyId == verifyingKey.keyId) {
            "signing and verifying key IDs must match"
        }
        require(signingKey.algorithm == verifyingKey.algorithm) {
            "signing and verifying algorithms must match"
        }
    }

    /** The key ID shared by both halves. */
    val keyId: String get() = signingKey.keyId

    /** The algorithm shared by both halves. */
    val algorithm: Algorithm get() = signingKey.algorithm
}
